tar/lib/src/reader.dart

```diff
@@ -0,0 +1,442 @@
+// Copyright 2020
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:chunked_stream/chunked_stream.dart';
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+
+import 'constants.dart';
+import 'exceptions.dart';
+import 'format.dart';
+import 'header.dart';
+import 'sparse_entry.dart';
+import 'stream.dart';
+import 'utils.dart';
+
+/// [TarReader] provides sequential access to the TAR files in a TAR archive.
+/// It is designed to read from a stream and to spit out substreams for
+/// individual file contents in order to minimize the amount of memory needed
+/// to read each archive where possible.
+@sealed
+class TarReader {
+  /// A chunked stream iterator to enable us to get our data 512 bytes at a time.
+  final ChunkedStreamIterator<int> _chunkedStream;
+
+  /// The [TarHeader] for the current file.
+  TarHeader header;
+
+  /// The current file contents.
+  Stream<List<int>> contents;
```

Comment:
We should probably make this a getter... so that users can't assign to this :D

And maybe we should document in the comment how we expect people to read this..
Do we always expect them to drain it? What if they current entry is a directory? Can this be `null`.

It might be fair to say that if the current entry has contents then this must be drained.. but that if the current entry doesn't have contents draining it unnecessary (but allowed). What is the behavior?

Reply:
Changed it to a getter!

Do you think I have to worry about the same issue as what you were commenting about `header`, that people while access it while `TarReader.next()` is still running?